packages/firebase-core/utils.ts

```diff
@@ -37,14 +37,15 @@ export function serialize(data: any, wrapPrimitives: boolean = false): any {
 				}
 
 				if (Array.isArray(data)) {
-					return NSArray.arrayWithArray((<any>data).map(serialize));
+					return NSArray.arrayWithArray(data.map((el) => serialize(el, wrapPrimitives)).filter((el) => el !== null && el !== NSNull.new()));
 				}
 
-				let node = {} as any;
-				Object.keys(data).forEach(function (key) {
-					let value = data[key];
-					node[key] = serialize(value, wrapPrimitives);
-				});
+				const node = Object.fromEntries(
+					Object.entries(data)
+						.map(([key, value]) => [key, serialize(value, wrapPrimitives)])
+						.filter(([key, value]) => value !== null && value !== NSNull.new())
```

Comment:
NSNull should be used for null values and maybe only filter undefined also `value !== NSNull.new()` will cause a new instance of NSNull to be created for every filter call. Something like `value instanceof NSNull` would be the way to go

Reply:
@triniwiz nice catch about creating NSNull instance!

What do you mean by filtering only undefined? The `serialize` doesn't have any code paths that returns undefined, it either returns `null` or `NSNull`. From my tests the error is thrown for both `null` and `undefined` passed inside objects or arrays

Reply:
Right silly meI forgot serialize doesn't return undefined, so some use cases require the nsnull object and some work fine with null it's the reason for the wrap primitive but as you mentioned the second param can become an index instead of the wrap primitive which can cause it to fail

Reply:
@triniwiz the NSNull check was indeed unnecessary. I updated the code to remove it but left the `null` check.

```ts
serialize({ aa: null, bb: null }, false); // filters out null
serialize([null, null], false); // filters out null

serialize({ aa: null, bb: null }, true); // converts null to NSNull
serialize([null, null], true); // converts null to NSNull

serialize({ aa: undefined, bb: undefined }, false); // filters out undefined
serialize([undefined, undefined], false); // filters out undefined

serialize({ aa: undefined, bb: undefined }, true); // converts undefined to NSNull
serialize([undefined, undefined], true); // converts undefined to NSNull
```